my/coding/commits.py

```diff
@@ -1,38 +1,64 @@
 """
 Git commits data for repositories on your filesystem
 """
+REQUIRES = [
+    'gitpython',
+]
+
 
 import shutil
 import string
 from pathlib import Path
 from datetime import datetime, timezone
-from typing import List, NamedTuple, Optional, Dict, Any, Iterator, Set
+from dataclasses import dataclass, field
+import functools
+from typing import List, NamedTuple, Optional, Iterator, Set, Sequence
+
+
+from my.core import PathIsh, LazyLogger, make_config
+from my.core.cachew import cache_dir
+from my.core.common import mcachew
+from my.core.warnings import high
+
+
+from my.config import commits as user_config
+@dataclass
+class commits_cfg(user_config):
+    roots: Sequence[PathIsh] = field(default_factory=list)
+    emails: Optional[Sequence[str]] = None
+    names: Optional[Sequence[str]] = None
 
-from ..core.common import PathIsh, LazyLogger, mcachew, Stats
-from ..core.cachew import cache_dir
-from ..core.warnings import high
 
-# TODO: create user_config dataclass?
-from my.config import commits as config
+# experiment to make it lazy?
+# would be nice to have a nicer syntax for it... maybe make_config could return a 'lazy' object
+@functools.lru_cache()
+def config() -> commits_cfg:
```

Comment:
Not sure about lru_cache in particular, but thinking that 'lazy' instantiation like this might be good to use in general, often the fact that config is constructed on the top level makes it harder to override/test.

Guess relevant to https://github.com/karlicoss/HPI/issues/46

Reply:
Hmm..

I'd think one might run into less problems if you avoid  `lru_cache` for things like config, because it means the user has less options to do things like monkey patching.

It does look/feel worse to have to handle caching the result of some computation after the config has been loaded. I think previously some of that had been done in `@property` functions on the config itself, but that doesn't feel like the best solution.

I think I agree with your comment:

> would be nice to have a nicer syntax for it... maybe make_config could return a 'lazy' object

Reply:
Another way that I've often handled this is to have an [`inputs` function](https://github.com/seanbreckenridge/HPI/blob/master/my/ttt.py#L39-L52) which takes what the config gives you, and computes the resulting inputs for each of the functions. Then that function is the [default parameter for any function](https://github.com/seanbreckenridge/HPI/blob/master/my/ttt.py#L70) which requires the input files

Is a bit confusing on first glance, but is nice for testing purposes and being able to specify a subset of the paths if I want to do something in the repl

Reply:
Hmm yeah ok, good point about caching!
I guess it doesn't hurt to instantiate the config multiple times, negligible in comparison with the rest of computation

Reply:
> the default parameter for any function which requires the input files

Yeah, I thought about that too, but then it often requires propagating the whole thing down the call stack.. I guess it's a classic 'state' problem